### DISCUSSION TITLE
vm.startBroadcast without arguments

### DISCUSSION THREAD

**Discussion by ducanh2706** [Cyfrin/foundry-full-course-cu#773] [Q&A] [answered]
I want to ask how the startBroadcast work in this case:

```solidity
// SPDX-License-Identifier: MIT

pragma solidity ^0.8.18;

import {Script, console} from "forge-std/Script.sol";
import {FundMe} from "../src/FundMe.sol";
import {HelpConfig} from "./HelpConfig.s.sol";

contract DeployFundMe is Script {
    function run() public returns (FundMe) {
        HelpConfig helpConfig = new HelpConfig();
        (address ethUsdPriceFeed) = helpConfig.activeNetworkConfig();
        console.log("Sender is %s", msg.sender);
        vm.startBroadcast();
        console.log("Sender is %s", msg.sender);
        FundMe fundMe = new FundMe(ethUsdPriceFeed);
        console.log("Owner is: %s", fundMe.getOwner());
        vm.stopBroadcast();
        return fundMe;
    }

}
```
The log is:
```
Sender is 0x7FA9385bE102ac3EAc297483Dd6233D62b3e1496
 Sender is 0x7FA9385bE102ac3EAc297483Dd6233D62b3e1496
 Owner is: 0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38
```

The owner is the msg.sender from the test call this script
```solidity
// SPDX-License-Identifier: MIT

pragma solidity ^0.8.18;

import {Test, console} from "forge-std/Test.sol";
import {FundMe} from "../../src/FundMe.sol";
import {DeployFundMe} from "../../script/DeployFundMe.s.sol";

contract FundMeTest is Test{
    FundMe fundMe;

    address public USER = makeAddr("DucAnhLe");
    uint256 constant public SEND_VALUE = 0.1 ether;
    uint256 constant public START_BALANCE = 100 ether;
    uint256 constant public GAS_PRICE = 1;
    address public owner;


    function setUp() external {
        owner = msg.sender;
        console.log("Initial sender: %s", msg.sender);
        DeployFundMe deployFundMe = new DeployFundMe();
        fundMe = deployFundMe.run();
        // fundMe = new FundMe(0x694AA1769357215DE4FAC081bf1f309aDC325306);
        vm.deal(USER, START_BALANCE);
    }
   }
   ```
   
   Why is the owner and the sender is different? I understand that the sender in the script contract is actually the FundMeTest contract address. Just wonder how the startBroadcast work inside.

**Comment by theirrationalone** [comment, marked as answer]
```log
Owner is: 0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38
```

`vm.broadcast()` is half truth, `vm.broadcast(<pk>)` is the full truth. 

If we won't provide any private key to `vm.broadcast()` explicitly. In this case, then, the foundry does that for us. Surprisingly, Foundry chooses one of fake accounts private key and put inside `vm.broadcast(here)` for us, implicitly. 


As a default need, Foundry spins a temporary anvil, Selects one account as a deployer to deploy each & every scripts on temporary anvil chain to facilitate deployments temporarily, from multiple fake accounts available to facilitate same. 

Note: we can't use `vm.broadcast()` and `vm.prank()` linearly. Indeed, the purpose of both is same but from different contexts, `vm.broadcast()` for scripts and `vm.prank()` for tests scripts. 

Therefore, Foundry ignores `msg.sender` if it finds either `vm.broadcast()` without `pk` Or `vm.broadcast(<pk>)` with private key, anywhere into the script. 

That's why, `msg.sender` isn't considered the deployer of `FundMe` and if `msg.sender` isn't the deployer then `msg.sender` can't be the `owner`, the owner is the one whose private-key was inserted into `vm.broadcast(here) ` by foundry. 

@ducanh2706 !, For fun sake, You can cross check the owner:

```log
Owner is: 0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38
```

Just spin the anvil, and you will find that `0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38` is one among provided by anvil.

**Reply by ducanh2706** [reply]
Wow, thanks for your detailed explanation. I think I understand it now.
Actually, when I spinned the anvil, I couldn't find any private-key/address that is similar to the private-key/address in the log. I think that private-key/address is just randomly generated, because when I console.log the msg.sender.balance, the result is:
```
79228162514264337593543950335
```
Please correct me if I said anything wrong ❤️

**Reply by theirrationalone** [reply]
Of course, You may right also... Cause that just randomly chosen by foundry.... Woww 🔥🔥 You got 101% + corrected me +1% , Thanks too you 😉, 

Glad to work with You @ducanh2706 ! 👏

**Reply by rocknet** [reply]
I know this is a bit later, but just to nail this down totally for me, it seems the main use case of seeting the pk in `startBroadcast` is when running tests with a forked chain, does this sound right?  The reason I'm trying to clarify is that, you can totally deploy contracts and send transactions using `vm.startBroadcast()`, without specifying the key, if you're doing something like `forge script --rpc-url $x --private-key $y`.  I'm thinking that the combination of supplying `rpc-url` and `private-key`, rather than `--fork-url`, is setting the pk specified, behind the scenes.

**Reply by theirrationalone** [reply]
> I know this is a bit later, but just to nail this down totally for me, it seems the main use case of seeting the pk in `startBroadcast` is when running tests with a forked chain, does this sound right?  The reason I'm trying to clarify is that, you can totally deploy contracts and send transactions using `vm.startBroadcast()`, without specifying the key, if you're doing something like `forge script --rpc-url $x --private-key $y`.  I'm thinking that the combination of supplying `rpc-url` and `private-key`, rather than `--fork-url`, is setting the pk specified, behind the scenes.
> 

I think Env * | inline cmds inj., takes precedence.
